File: VB/GridMVVMBindableColumns/ViewModel.vb
Imports System.Collections.ObjectModel
Imports System.Windows.Input
Imports DevExpress.Xpf.Core.Commands
Imports System.ComponentModel

Namespace Model

    Public Class ViewModel
        Implements INotifyPropertyChanged

        Private _Source As ObservableCollection(Of Model.DataItem), _Categories As ObservableCollection(Of Model.Category)

        Private categoryColumnVisibilityField As Boolean = True

        Public Property Source As ObservableCollection(Of DataItem)
            Get
                Return _Source
            End Get

            Private Set(ByVal value As ObservableCollection(Of DataItem))
                _Source = value
            End Set
        End Property

        Public Property Categories As ObservableCollection(Of Category)
            Get
                Return _Categories
            End Get

            Private Set(ByVal value As ObservableCollection(Of Category))
                _Categories = value
            End Set
        End Property

        Public Property CategoryColumnVisibility As Boolean
            Get
                Return categoryColumnVisibilityField
            End Get

            Set(ByVal value As Boolean)
                If categoryColumnVisibilityField = value Then Return
                categoryColumnVisibilityField = value
                RaisePropertyChanged("CategoryColumnVisibility")
            End Set
        End Property

        Public Property ShowHideCategoriesColumn As ICommand

        Public Sub New()
            Source = DataItem.Data
            Categories = New ObservableCollection(Of Category)() From {New Category() With {.Id = 0, .Name = "A category"}, New Category() With {.Id = 1, .Name = "B category"}, New Category() With {.Id = 2, .Name = "C category"}, New Category() With {.Id = 3, .Name = "D category"}}
            ShowHideCategoriesColumn = New DelegateCommand(Of Object)(Sub(obj) CategoryColumnVisibility = Not CategoryColumnVisibility, Nothing)
        End Sub

        Private Sub RaisePropertyChanged(ByVal name As String)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(name))
        End Sub

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged
    End Class

    Public Class Category

        Public Property Id As Integer

        Public Property Name As String
    End Class

    Public Class DataItem

        Public Shared ReadOnly Property Data As ObservableCollection(Of DataItem)
            Get
                Dim res As ObservableCollection(Of DataItem) = New ObservableCollection(Of DataItem)()
                For i As Integer = 0 To 100 - 1
                    res.Add(New DataItem() With {.Id = i, .Name = "Name" & i.ToString(), .Category = i Mod 4})
                Next

                Return res
            End Get
        End Property

        Public Property Id As Integer

        Public Property Name As String

        Public Property Category As Integer

        Public Overrides Function ToString() As String
            Return Name
        End Function
    End Class
End Namespace

File: CS/GridMVVMBindableColumns/ViewModel.cs
using System;
using System.Collections.ObjectModel;
using System.Windows.Input;
using DevExpress.Utils.Commands;
using DevExpress.Xpf.Core.Commands;
using System.Collections;
using System.Linq;
using System.ComponentModel;

namespace Model {
    public class ViewModel : INotifyPropertyChanged {
        bool categoryColumnVisibility = true;
        public ObservableCollection<DataItem> Source { get; private set; }
        public ObservableCollection<Category> Categories { get; private set; }
        public bool CategoryColumnVisibility {
            get { return categoryColumnVisibility; }
            set {
                if(categoryColumnVisibility == value)
                    return;
                categoryColumnVisibility = value;
                RaisePropertyChanged("CategoryColumnVisibility");
            }
        }
        public ICommand ShowHideCategoriesColumn { get; set; }
        public ViewModel() {
            Source = DataItem.Data;
            Categories = new ObservableCollection<Category>() {
                new Category() { Id=0, Name="A category"},
                new Category() { Id=1, Name="B category"},
                new Category() { Id=2, Name="C category"},
                new Category() { Id=3, Name="D category"}
            };
            ShowHideCategoriesColumn = new DelegateCommand<object>(obj => {
                CategoryColumnVisibility = !CategoryColumnVisibility;
            }, null);
        }

        void RaisePropertyChanged(string name) {
            if(PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs(name));
        }

        public event PropertyChangedEventHandler PropertyChanged;
    }

    public class Category {
        public int Id { get; set; }
        public string Name { get; set; }
    }
    public class DataItem {
        public static ObservableCollection<DataItem> Data {
            get {
                ObservableCollection<DataItem> res = new ObservableCollection<DataItem>();
                for(int i = 0; i < 100; i++) {
                    res.Add(new DataItem() { Id = i, Name = "Name" + i.ToString(), Category = i % 4 });
                }
                return res;
            }
        }
        public int Id { get; set; }
        public string Name { get; set; }
        public int Category { get; set; }
        public override string ToString() {
            return Name;
        }
    }
}
